### ISSUE TITLE
Odd behavior parsing floats in JSON doc

### ISSUE THREAD

**Issue by dvdplm** [serde-rs/json#378]
When parsing a big JSON document with color conversion data I came across two cases where serde does not deserialise the data as I would have expected it. It may very well be I'm not aware of how floats work and that I'm missing something fundamental. When serde parse 98.8833954570195**3** I get 98.8833954570195**2** back.

Example code where the assertion fails:

```rust
#[macro_use]
extern crate serde_derive;
extern crate serde;
extern crate serde_json;

use std::collections::HashMap;

#[derive(Deserialize, Debug)]
struct AStruct {
    afloat: f64,
}  
fn main() {
    let raw_data = r#"{"val":{"afloat":98.88339545701953}}"#;
    println!("Raw JSON data: {:?}", raw_data);
    let floats: HashMap<String, AStruct> = serde_json::from_str(&raw_data).unwrap();
    println!("Parsed JSON data: {:?}", floats);
    assert!(floats["val"].afloat == 98.88339545701953);
}
```

I suspect this is related to https://github.com/serde-rs/json/issues/128 but I'm not sure if I fully understand the problem there.

**Comment by dtolnay** [comment]
This is a consequence of #128. The floating point value closest to `"98.88339545701953"` has an exact value of the top line, whereas the one we parse is the bottom line.

```rust
98.883395457019531704645487479865550994873046875  // closest float above
98.88339545701953                                 // exact value (not a float)
98.8833954570195174937907722778618335723876953125 // closest float below
```

Correct parsing is tricky because it would need to be able to tell the difference between these numbers that differ by 10^-48 which is about 2^-160, way outside the range of 64-bit types.

```rust
fn main() {
    println!("{}", 98.883395457019524599218129878863692283630371093750f64);
    println!("{}", 98.883395457019524599218129878863692283630371093749f64);
}
```

**Comment by dvdplm** [comment]
I think I understand, let me try to rephrase it to see if I got it right: 98.88339545701953 is close to the limit of what serde (and rust) can represent with f64 and the parsing technique used has to pick one of two possible numbers (closest larger and closest smaller) to approximate the input. Sometimes it picks the wrong one. Correct so far?

Questions:
1. are there any plans to change the behavior here? Is there even a better way to do this?
2. on #128 you mention serde is creating a u64, cast it to f64 and divides by 10e14 – why 14? Skimming through the V8 code at https://github.com/google/double-conversion/blob/master/double-conversion/strtod.cc I see they mention 15 significant digits in a double (2^53).

Sorry for the silly questions, my understanding of floats is hazy!

**Comment by dtolnay** [comment]
- There is no 64-bit float that is equal to that number. We do a fast, simple, and reasonably accurate calculation to find a float that is very close to that number. Finding the float that is closer to that number than any other float is slow and complicated.
1. #128 is tracking our intention to fix the behavior.
2. The divisor depends on number of digits after the decimal point, so `1.234` is `1234 / 10^3`.

**Comment by dvdplm** [comment]
Closing in favour of #128
